Removed RigidBodyTree constructor that takes as input a URDF file name.

liangfok opened the pull request:
This goes towards the goal of decoupling the `RigidBodyTree` from the parsers. The layering should be: parsers depend on `RigidBodyTree` and _not the other way around_.

I was motivated to prioritize this because it will make it easier for me to remove the "directory crawling" that's currently being done by `populatePackageMap()`, which is defined in `xml_util.h`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4317)
<!-- Reviewable:end -->

liangfok commented:
+@amcastro-tri for feature review please.
+@sammy-tri for platform review please.

Thanks!

---

Review status: 0 of 28 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-:-KXqxJLvHv8p6aN9XOia:b-gmg4pt)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Oops. I broke some unit tests. Applying the "do not review" tag while I fix them.

---

Review status: 0 of 28 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-:-KXr0zYD76c0uCdnaKAR:b-n3tk8b)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
> The layering should be: parsers depend on RigidBodyTree and not the other way around.

I don't agree with this strategy, @liangfok. A parser's job is to convert a serialized representation to an in-memory representation. Then classes like RBTree, RBPlant, Sensors, GeometryWorld or whatever can construct themselves from the in-memory representation by extracting information they care about (e.g. bodies and joints for RBTree, lidar parameters for a sensor, and maybe roads and traffic lights for something else). It wouldn't make sense for a parser to have to learn about all the things that might need to be constructed from the parsed information, so IMO it shouldn't start by learning how to build an RBTree. 

Where I would like us to go is more like
```c++
in_memory_rep = Parse("file.urdf"); // or Parse("file.sdf"); same in-memory representation
RigidBodyTree my_tree(in_memory_rep);
Sensors my_sensors(in_memory_rep);
// ...
```

---

Review status: 0 of 34 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-:-KXseR3Yrv-VlolHHz8e:b-am9l5h)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@sherm1, I agree with everything you've said, but I believe the first step towards achieving your (long term) vision is to modify `RigidBodyTree` to not call the parsers, which this PR takes a step towards doing. After this PR, follow-up PRs will (1) remove the deprecated `RigidBodyTree::addRobotFromFoo()` methods, and then (2) pull the parsers into their own libraries. Longer term, we can modify the parsers to output an Intermediate Representation (IR), and then add constructors to `RigidBodyTree` and the other modeling objects that accept the IR as a parameter, enabling them to construct themselves based on the IR.

I'd also like to recall our group meeting from October 21, 2016 where I [presented some slides](https://docs.google.com/presentation/d/1wBaxrB3umlhE9vAPsy59uO3TPI271GJxCKK8RuHokWw/edit#slide=id.g138a6a73da_0_13) containing ideas on how to decouple the parsers from the `RigidBodyTree`. Originally, I believe my slides contained your exact vision, which I replicate​ below.

![Screen Shot 2016-12-01 at 6.31.01 AM.png](https://files.reviewableusercontent.io/images/reviewable/1388098/zdhiCpjFMSl0cMLZBAcr5aIn/Screen%20Shot%202016-12-01%20at%206.31.01%20AM.png)

However, during the meeting, questions were raised about why `RigidBodyTree` and other existing modeling elements couldn't serve as the IR. After the group meeting, @david-german-tri helped me design what I consider to be an intermediate step towards your vision. In particular, see slide 4, a snapshot of which I replicate below.


![Screen Shot 2016-12-01 at 6.20.35 AM.png](https://files.reviewableusercontent.io/images/reviewable/1388098/thYl1C__sjZus6pm5jI_TVOL/Screen%20Shot%202016-12-01%20at%206.20.35%20AM.png)

I'm going to add you as a platform reviewer just to ensure this PR does not go in without your blessings.
+@sherm1 for platform review please.

---

Review status: 0 of 34 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-:-KXu3XoEaarHlio3jJOn:bgevw7y)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 2 of 34 files at r1.
Review status: 2 of 34 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
First pass.
I agree with both @sherm1 and @liangfok on the discussion above. However @sherm1 what you are suggesting is not even possible right now. Right now our intermediate represenation for `RBT` is `RBT<double>`, period. I think the path @liangfok is suggesting makes sense and this PR does move in the right direction, which is taking all the parsing code out of RBT and having outside methods (like `AddModelInstanceFromUrdfFileToWorld` so heavily used in this PR) to do the parsing.
The workflow this PR proposes is:
```c++
RigidBodyTree<double> model;
drake::parsers::AddModelInstanceFromUrdfFileToWorld("model.urdf", &model);
```
One issue to resolve in this workflow is the later instantiation of Bullet elements per #2661 and also mentioned in #4204 but that can be resolved with a wrapper around `CollisionElements` with a PIMPL implementation to `BulletCollisionElement`'s that get instantiated later, say with `RBT::compile()`.

I think we all agree on the need of a IR but that would take longer to achieve and this PR I think does help in taking the parsing out of RBT.

---

Reviewed 9 of 34 files at r1.
Review status: 9 of 34 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/bindings/swig/rbtree.i, line 82 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXurx4u1d743LUXQUQg:-KXurx4u1d743LUXQUQh:b26uqgw) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/bindings/swig/rbtree.i#L82)):*
> ```
> %include "drake/multibody/joints/floating_base_types.h"
> %extend RigidBodyTree {
>   RigidBodyTree(const std::string& urdf_filename, const std::string& joint_type) {
> ```

I would provide here the default API which would be having `joint_type = "ROLLPITCHYAW"`  by default. That way you dont need to change the python examples at all. Also I think this changes the Matlab API if they use this constructor.

---

*[drake/examples/Acrobot/test/acrobot_urdf_dynamics_test.cc, line 8 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuqDe3jl8KtpXf6b_t:-KXuqDe3jl8KtpXf6b_u:bitlabg) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/Acrobot/test/acrobot_urdf_dynamics_test.cc#L8)):*
> ```cpp
> #include "drake/common/eigen_matrix_compare.h"
> #include "drake/examples/Acrobot/acrobot_plant.h"
> #include "drake/multibody/joints/floating_base_types.h"
> ```

is `floating_base_types.h` needed? and elsewhere.

---

*[drake/examples/kuka_iiwa_arm/kuka_ik_demo.cc, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXupaE_jOt76iiCvDtH:-KXupaEa6e3Kr-Hn9WQa:b-ji1gxx) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/kuka_iiwa_arm/kuka_ik_demo.cc#L17)):*
> ```cpp
> #include "drake/multibody/constraint/rigid_body_constraint.h"
> #include "drake/multibody/ik_options.h"
> #include "drake/multibody/joints/floating_base_types.h"
> ```

is `floating_base_types.h` needed?

---

*[drake/examples/QPInverseDynamicsForHumanoids/test/lcm_utils_test.cc, line 11 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuqQL--W7yt3ocgzYU:-KXuqQL0_8SvX0L9Y4bJ:b-ji1gxx) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/QPInverseDynamicsForHumanoids/test/lcm_utils_test.cc#L11)):*
> ```cpp
> #include "drake/examples/QPInverseDynamicsForHumanoids/example_qp_input_for_valkyrie.h"
> #include "drake/examples/QPInverseDynamicsForHumanoids/lcm_utils.h"
> #include "drake/multibody/joints/floating_base_types.h"
> ```

is `floating_base_types.h` needed?

---

*[drake/examples/Valkyrie/test/valkyrie_ik_test.cc, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXupUf-ZiKC1yuBRZfO:-KXupUf0B_onnqppUzGc:b-g4ivo) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/Valkyrie/test/valkyrie_ik_test.cc#L21)):*
> ```cpp
> #include "drake/systems/framework/diagram_builder.h"
> #include "drake/systems/framework/primitives/constant_vector_source.h"
> #include "drake/multibody/joints/floating_base_types.h"
> ```

Why this header is needed now?

---

*[drake/examples/Valkyrie/test/valkyrie_ik_test.cc, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXupCSwfW9GaJxT4FJ_:-KXupCSxWXw4bld5Wt58:b9bp3z0) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/Valkyrie/test/valkyrie_ik_test.cc#L70)):*
> ```cpp
> 
> GTEST_TEST(ValkyrieIK_Test, ValkyrieIK_Test_StandingPose_Test) {
>   auto tree = std::make_shared<RigidBodyTree<double>>();
> ```

No shared pointers in Drake anymore! new code should not use `shared_ptr` at all. Use `unique_ptr` please.

---

*[drake/multibody/rigid_body_tree.cc, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuoe7b6-Besp7O-JOF:-KXuoe7b6-Besp7O-JOG:bq6s2ev) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/multibody/rigid_body_tree.cc#L23)):*
> ```cpp
> #include "drake/multibody/parser_sdf.h"
> #include "drake/multibody/parser_urdf.h"
> #include "drake/multibody/xml_util.h"
> ```

why this new header now?

---

*[drake/multibody/rigid_body_tree.h, line 89 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuo1rIgxjh3-l8Pq52:-KXuo1rJh4fX8z1FiVXZ:bnopzig) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/multibody/rigid_body_tree.h#L89)):*
> ```objc
>   /**
>    * A constructor that initializes the gravity vector to be [0, 0, -9.81] and
>    * creates an empty world.
> ```

Not entirely empty. Update to mention there actually is one body (i.e. `RBT::get_num_bodies()` returns 1)named "world" which the user can request with `RBT::world()`.

---

*[drake/systems/controllers/InstantaneousQPController.h, line 10 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXusT94yk97JrsgpLqN:-KXusT94yk97JrsgpLqO:b-ji1gxx) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/systems/controllers/InstantaneousQPController.h#L10)):*
> ```objc
> #include "drake/solvers/gurobi_qp.h"
> #include "drake/lcmt_qp_controller_input.hpp"
> #include "drake/multibody/joints/floating_base_types.h"
> ```

is `floating_base_types.h` needed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-:-KXus_NvHPwe7AOBBroy:b-mylar7)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@amcastro-tri I addressed your comments. PTAL. Thanks.

---

Review status: 9 of 34 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/bindings/swig/rbtree.i, line 82 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXurx4u1d743LUXQUQg:-KXv-GbZyOV0IGPt9txv:b-qu6n94) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/bindings/swig/rbtree.i#L82)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I would provide here the default API which would be having `joint_type = "ROLLPITCHYAW"`  by default. That way you dont need to change the python examples at all. Also I think this changes the Matlab API if they use this constructor.
</blockquote></details>

Good point. Done.

---

*[drake/examples/Acrobot/test/acrobot_urdf_dynamics_test.cc, line 8 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuqDe3jl8KtpXf6b_t:-KXv-Q2Kul91nZljEWBq:buyphak) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/Acrobot/test/acrobot_urdf_dynamics_test.cc#L8)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

is `floating_base_types.h` needed? and elsewhere.
</blockquote></details>

Yes, note the `multibody::joints::kFixed` parameter below. This is simply following the "Include What You Use" (IWYU) style guide rule.

---

*[drake/examples/kuka_iiwa_arm/kuka_ik_demo.cc, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXupaE_jOt76iiCvDtH:-KXv-eTOfSPGtA_C63ec:b6q47h7) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/kuka_iiwa_arm/kuka_ik_demo.cc#L17)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

is `floating_base_types.h` needed?
</blockquote></details>

Ditto except this file is referencing `multibody::joints::kFixed`.

---

*[drake/examples/QPInverseDynamicsForHumanoids/test/lcm_utils_test.cc, line 11 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuqQL--W7yt3ocgzYU:-KXv-mfQonbhtG3MmJz9:b3lsgqa) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/QPInverseDynamicsForHumanoids/test/lcm_utils_test.cc#L11)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

is `floating_base_types.h` needed?
</blockquote></details>

This file references ` multibody::joints::kRollPitchYaw`.

---

*[drake/examples/Valkyrie/test/valkyrie_ik_test.cc, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXupUf-ZiKC1yuBRZfO:-KXv-t0jpPdr2dG0RXGI:b-5fl7w0) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/Valkyrie/test/valkyrie_ik_test.cc#L21)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Why this header is needed now?
</blockquote></details>

This file references `multibody::joints::kRollPitchYaw`.

---

*[drake/examples/Valkyrie/test/valkyrie_ik_test.cc, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXupCSwfW9GaJxT4FJ_:-KXv06-SBewJuoT60S9e:b8htqjj) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/Valkyrie/test/valkyrie_ik_test.cc#L70)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

No shared pointers in Drake anymore! new code should not use `shared_ptr` at all. Use `unique_ptr` please.
</blockquote></details>

Done. This was a holdover from the original code. I agree it can be fixed in this PR.

---

*[drake/multibody/rigid_body_tree.cc, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuoe7b6-Besp7O-JOF:-KXv0Rz7coXeQt8UrQrv:bmzrc44) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/multibody/rigid_body_tree.cc#L23)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

why this new header now?
</blockquote></details>

This file references `PackageMap`, which is defined in `xml_util.h`. I added a TODO comment to remind myself to remove it once `PackageMap` is no longer referenced in this file. It should only be referenced by the parsers.

---

*[drake/multibody/rigid_body_tree.h, line 89 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuo1rIgxjh3-l8Pq52:-KXv18hVJSCsHeWNE6oA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/multibody/rigid_body_tree.h#L89)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Not entirely empty. Update to mention there actually is one body (i.e. `RBT::get_num_bodies()` returns 1)named "world" which the user can request with `RBT::world()`.
</blockquote></details>

Done.

---

*[drake/systems/controllers/InstantaneousQPController.h, line 10 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXusT94yk97JrsgpLqN:-KXv1OES8ap0QXSUCzYP:b50735t) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/systems/controllers/InstantaneousQPController.h#L10)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

is `floating_base_types.h` needed?
</blockquote></details>

Yes, this file is now referencing `multibody::joints::kRollPitchYaw`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-:-KXv1TeecaG50kowjshn:b-28jvfw)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Reviewed 4 of 34 files at r1, 1 of 9 files at r2.
Review status: 9 of 37 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[drake/bindings/python/pydrake/test/testPR2IK.py, line 9 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGYTl8u8xs6c6VXkm:-KXvGYTl8u8xs6c6VXkn:b-d1w4ug) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testPR2IK.py#L9)):*
> ```Python
> robot = pydrake.rbtree.RigidBodyTree(os.path.join(pydrake.getDrakePath(),
>                                                   "examples/PR2/pr2.urdf"),
>                                      "ROLLPITCHYAW")
> ```

Extra argument no needed anymore?

---

*[drake/bindings/python/pydrake/test/testRBTCoM.py, line 12 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGWrlDf6ION7TsqXS:-KXvGWrlDf6ION7TsqXT:b-d1w4ug) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTCoM.py#L12)):*
> ```Python
>         r = pydrake.rbtree.RigidBodyTree(os.path.join(pydrake.getDrakePath(),
>                                              "examples/Pendulum/Pendulum.urdf"),
>                                         "ROLLPITCHYAW")
> ```

Extra argument no needed anymore?

---

*[drake/bindings/python/pydrake/test/testRBTCoM.py, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGXDpvnQkBDg_7pu6:-KXvGXDpvnQkBDg_7pu7:b-d1w4ug) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTCoM.py#L23)):*
> ```Python
>         r = pydrake.rbtree.RigidBodyTree(os.path.join(pydrake.getDrakePath(),
>                                              "examples/Pendulum/Pendulum.urdf"),
>                                          "ROLLPITCHYAW")
> ```

Extra argument no needed anymore?

---

*[drake/bindings/python/pydrake/test/testRBTIK.py, line 11 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGSJsY-P8nIxn_iIf:-KXvGSJt-ssk5q9sEEd9:b-d1w4ug) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTIK.py#L11)):*
> ```Python
>         r = pydrake.rbtree.RigidBodyTree(
>             os.path.join(pydrake.getDrakePath(),
>                          "examples/Pendulum/Pendulum.urdf"), "ROLLPITCHYAW")
> ```

Extra argument no needed anymore?

---

*[drake/bindings/python/pydrake/test/testRBTTransformPoints.py, line 12 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGczaaZJ2QeFIFyTG:-KXvGczaaZJ2QeFIFyTH:b-d1w4ug) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTTransformPoints.py#L12)):*
> ```Python
>         r = pydrake.rbtree.RigidBodyTree(os.path.join(pydrake.getDrakePath(), "examples/Pendulum/Pendulum.urdf"),
>                                          "ROLLPITCHYAW")
>         self.assertEqual(r.number_of_positions(), 7)
> ```

Extra argument no needed anymore?

---

*[drake/bindings/python/pydrake/test/testRBTTransformPoints.py, line 22 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGdeT4JvpRUDv9tK3:-KXvGdeT4JvpRUDv9tK4:b-d1w4ug) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTTransformPoints.py#L22)):*
> ```Python
>     def test_gradient(self):
>         r = pydrake.rbtree.RigidBodyTree(os.path.join(pydrake.getDrakePath(), "examples/Pendulum/Pendulum.urdf"),
>                                          "ROLLPITCHYAW")
> ```

Extra argument no needed anymore?

---

*[drake/bindings/python/pydrake/test/testRBTTransformPoints.py, line 38 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGe2WH0DPjflELV2K:-KXvGe2WH0DPjflELV2L:b-d1w4ug) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTTransformPoints.py#L38)):*
> ```Python
>     def test_big_gradient(self):
>         r = pydrake.rbtree.RigidBodyTree(os.path.join(pydrake.getDrakePath(), "examples/Pendulum/Pendulum.urdf"),
>                                          "ROLLPITCHYAW")
> ```

Extra argument no needed anymore?

---

*[drake/examples/Acrobot/test/acrobot_urdf_dynamics_test.cc, line 8 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuqDe3jl8KtpXf6b_t:-KXvFQFEeWuVePTurIBx:b-83pp) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/Acrobot/test/acrobot_urdf_dynamics_test.cc#L8)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Yes, note the `multibody::joints::kFixed` parameter below. This is simply following the "Include What You Use" (IWYU) style guide rule.
</blockquote></details>

Ah! I was confused because it wasn't there before. Is this a C++ style guide convention? 

---

*[drake/examples/Valkyrie/test/valkyrie_ik_test.cc, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXupCSwfW9GaJxT4FJ_:-KXvFvUnEPkfEjKQwAJR:bwzg45o) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/Valkyrie/test/valkyrie_ik_test.cc#L70)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Done. This was a holdover from the original code. I agree it can be fixed in this PR.
</blockquote></details>

I know. Thanks for fixing.

---

*[drake/multibody/rigid_body_tree.h, line 89 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuo1rIgxjh3-l8Pq52:-KXvGBbbL2T-CLjgUTGO:buz7eg1) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/multibody/rigid_body_tree.h#L89)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks. I am also adopting the `///` style for Doxygen in RBT (though I still didn't update the old ones). Could you change to that?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 5 of 35 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/bindings/python/pydrake/test/testRBTCoM.py, line 12 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGWrlDf6ION7TsqXS:-KXvdaObpI2sCJPcdcs5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTCoM.py#L12)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Extra argument no needed anymore?
</blockquote></details>

Done.

---

*[drake/bindings/python/pydrake/test/testRBTCoM.py, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGXDpvnQkBDg_7pu6:-KXvddsDTjg0D0K9Equj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTCoM.py#L23)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Extra argument no needed anymore?
</blockquote></details>

Done.

---

*[drake/examples/Acrobot/test/acrobot_urdf_dynamics_test.cc, line 8 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuqDe3jl8KtpXf6b_t:-KXveGkzLvJy6iSNsCC4:b-c5zdew) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/Acrobot/test/acrobot_urdf_dynamics_test.cc#L8)):*
Yes, here is the official text from the guide:

> You should include all the headers that define the symbols you rely upon, except in the unusual case of forward declaration. If you rely on symbols from bar.h, don't count on the fact that you included foo.h which (currently) includes bar.h: include bar.h yourself, unless foo.h explicitly demonstrates its intent to provide you the symbols of bar.h. However, any includes present in the related header do not need to be included again in the related cc (i.e., foo.cc can rely on foo.h's includes).

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

Drake clarifies this rule as follows:

> While we encourage you to Include What You Use (IWYU) since it improves code transparency and readability, it will not be strictly enforced. Instead, we enforce a “weak include” style that simply requires every symbol referenced within a file be covered by the transitive closure of all #include statements in the file. We decided to make this exception because (1) we can always adopt an IWYU rule later, (2) to reduce verbosity, and (3) we don’t have a tool to enforce IWYU at this time.

http://drake.mit.edu/code_style_guide.html#code-style-guide-cpp-exceptions

---

*[drake/examples/Valkyrie/test/valkyrie_ik_test.cc, line 70 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXupCSwfW9GaJxT4FJ_:-KXveii0rsaj8R80nccT:blyphgc) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/Valkyrie/test/valkyrie_ik_test.cc#L70)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I know. Thanks for fixing.
</blockquote></details>

No problem!

---

*[drake/multibody/rigid_body_tree.h, line 89 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuo1rIgxjh3-l8Pq52:-KXvettF2cir9TvYPZxk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/multibody/rigid_body_tree.h#L89)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Thanks. I am also adopting the `///` style for Doxygen in RBT (though I still didn't update the old ones). Could you change to that?
</blockquote></details>

Done.

---

*[drake/bindings/python/pydrake/test/testPR2IK.py, line 9 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGYTl8u8xs6c6VXkm:-KXvdMz6WSjfp5PTmJh3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testPR2IK.py#L9)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Extra argument no needed anymore?
</blockquote></details>

Done.

---

*[drake/bindings/python/pydrake/test/testRBTIK.py, line 11 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGSJsY-P8nIxn_iIf:-KXvdhlmAuw2b6mtOz-6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTIK.py#L11)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Extra argument no needed anymore?
</blockquote></details>

Done.

---

*[drake/bindings/python/pydrake/test/testRBTTransformPoints.py, line 12 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGczaaZJ2QeFIFyTG:-KXvdncKihdYl_1JsKDE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTTransformPoints.py#L12)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Extra argument no needed anymore?
</blockquote></details>

Done.

---

*[drake/bindings/python/pydrake/test/testRBTTransformPoints.py, line 22 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGdeT4JvpRUDv9tK3:-KXvdp3eNMjuw6dfqK7B:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTTransformPoints.py#L22)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Extra argument no needed anymore?
</blockquote></details>

Done.

---

*[drake/bindings/python/pydrake/test/testRBTTransformPoints.py, line 38 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXvGe2WH0DPjflELV2K:-KXvdqLzWM9Du5N1gjJ4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8ab8c08dd3d8493cf23fb6fd0435ddc082545845/drake/bindings/python/pydrake/test/testRBTTransformPoints.py#L38)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Extra argument no needed anymore?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Thanks, @liangfok and @amcastro-tri for the detailed explanations. Sounds great and I withdraw my objection. Thanks for including me in review, Liang. I'm confident it is in good hands so I'll -@sherm1. BTW, the "do not review" flag is still set.

---

Review status: 5 of 38 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-:-KXwVYd4GsjH8dbsmgfc:bu66nae)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Thanks @sherm1. I'll remove the "do not review" tag until after I get all of the pre-merge CI tests to pass.

---

Review status: 5 of 38 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-:-KXwwltALbMfjeYmd4O7:b-z2fns8)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Hmm, `linux-gcc-experimental-open-source` appears to have stalled; there are no "Details" link even after 3 hours.

@drake-jenkins-bot linux-gcc-experimental-open-source  please.

liangfok commented:
This PR passed all pre-merge CI tests and is thus ready for review.

liangfok commented:
I added a "high priority" label because this is blocking work on removing the directory crawling that's taking place within the parsers (resolution of #3435).

liangfok commented:
I suspect the linux-gcc-experimental-matlab tests failed b/c of #2410.
@drake-jenkins-bot linux-gcc-experimental-matlab please.

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 21 of 34 files at r1, 5 of 9 files at r2, 4 of 12 files at r3, 6 of 7 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-:-KY0Q_CNn6C0wghTlWVJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot  linux-clang-ninja-experimental-minimal please.

liangfok commented:
@drake-jenkins-bot linux-gcc-experimental-ros  please

liangfok commented:
@drake-jenkins-bot linux-gcc-ninja-experimental-open-source-debug please.

amcastro-tri commented:
Reviewed 1 of 7 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 17 of 34 files at r1, 4 of 9 files at r2, 7 of 12 files at r3, 6 of 7 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Acrobot/test/acrobot_urdf_dynamics_test.cc, line 8 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KXuqDe3jl8KtpXf6b_t:-KY0d0_MTyYT4bL0Ukd_:b-ucfknh) ([raw file](https://github.com/robotlocomotion/drake/blob/444e642d76870f1fe11d9abb86f52a8da5ee7c2f/drake/examples/Acrobot/test/acrobot_urdf_dynamics_test.cc#L8)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Yes, here is the official text from the guide:

> You should include all the headers that define the symbols you rely upon, except in the unusual case of forward declaration. If you rely on symbols from bar.h, don't count on the fact that you included foo.h which (currently) includes bar.h: include bar.h yourself, unless foo.h explicitly demonstrates its intent to provide you the symbols of bar.h. However, any includes present in the related header do not need to be included again in the related cc (i.e., foo.cc can rely on foo.h's includes).

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

Drake clarifies this rule as follows:

> While we encourage you to Include What You Use (IWYU) since it improves code transparency and readability, it will not be strictly enforced. Instead, we enforce a “weak include” style that simply requires every symbol referenced within a file be covered by the transitive closure of all #include statements in the file. We decided to make this exception because (1) we can always adopt an IWYU rule later, (2) to reduce verbosity, and (3) we don’t have a tool to enforce IWYU at this time.

http://drake.mit.edu/code_style_guide.html#code-style-guide-cpp-exceptions
</blockquote></details>

Thanks!

---

*[drake/multibody/constraint/test/rigid_body_constraint_test.cc, line 33 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-KY0fqCf6Iw5Lm1enNMT:-KY0fqCf6Iw5Lm1enNMU:b-4v4hbk) ([raw file](https://github.com/robotlocomotion/drake/blob/8db482efadd47f76047a37e0632bcb92ad9ce892/drake/multibody/constraint/test/rigid_body_constraint_test.cc#L33)):*
> ```cpp
>   kc1_lb << 0, 0, 0;
>   kc1_ub << 0, 0, 0;
>   std::unique_ptr<WorldCoMConstraint> world_com_constraint;
> ```

BTW, nice!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4317#-:-KY0gltq3idRYmMHemE9:bnfp4nl)*
<!-- Sent from Reviewable.io -->
